sdk/trace/src/main/scala/org/typelevel/otel4s/sdk/trace/data/EventData.scala

```diff
@@ -0,0 +1,146 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.typelevel.otel4s.sdk
+package trace.data
+
+import cats.Hash
+import cats.Show
+import cats.syntax.show._
+import org.typelevel.otel4s.Attribute
+import org.typelevel.otel4s.semconv.trace.attributes.SemanticAttributes
+
+import java.io.PrintWriter
+import java.io.StringWriter
+import scala.concurrent.duration.FiniteDuration
+
+/** Data representation of an event.
+  *
+  * @see
+  *   [[https://opentelemetry.io/docs/specs/otel/trace/api/#add-events]]
+  */
+sealed trait EventData {
+
+  /** The name of the event.
+    */
+  def name: String
+
+  /** The epoch time of the event.
+    */
+  def epoch: FiniteDuration
+
+  /** The attributes of the event.
+    */
+  def attributes: Attributes
+
+  override final def hashCode(): Int =
+    Hash[EventData].hash(this)
+
+  override final def equals(obj: Any): Boolean =
+    obj match {
+      case other: EventData => Hash[EventData].eqv(this, other)
+      case _                => false
+    }
+
+  override final def toString: String =
+    Show[EventData].show(this)
+}
+
+object EventData {
+  private final val ExceptionEventName = "exception"
+
+  /** Creates [[EventData]] with the given arguments.
+    *
+    * @param name
+    *   the name of the event
+    *
+    * @param epoch
+    *   the epoch time of the event
+    *
+    * @param attributes
+    *   the attributes to associate with the event
+    */
+  def apply(
+      name: String,
+      epoch: FiniteDuration,
+      attributes: Attributes
+  ): EventData =
+    Impl(name, epoch, attributes)
+
+  /** Creates [[EventData]] from the given exception.
+    *
+    * The name of the even will be `exception`.
+    *
+    * Exception details (name, message, and stacktrace) will be added to the
+    * attributes.
+    *
+    * @param epoch
+    *   the epoch time of the event
+    *
+    * @param exception
+    *   the exception to associate with the event
+    *
+    * @param attributes
+    *   the attributes to associate with the event
+    */
+  def fromException(
+      epoch: FiniteDuration,
+      exception: Throwable,
+      attributes: Attributes
+  ): EventData = {
+    val allAttributes = {
+      val builder = Vector.newBuilder[Attribute[_]]
+
+      builder.addOne(
+        Attribute(SemanticAttributes.ExceptionType, exception.getClass.getName)
+      )
+
+      val message = exception.getMessage
+      if (message != null) {
+        builder.addOne(Attribute(SemanticAttributes.ExceptionMessage, message))
+      }
+
+      val stringWriter = new StringWriter()
+      val printWriter = new PrintWriter(stringWriter)
+      exception.printStackTrace(printWriter)
+
+      builder.addOne(
+        Attribute(SemanticAttributes.ExceptionStacktrace, stringWriter.toString)
+      )
```

Comment:
Some exceptions have no stack traces e.g. `NoStackTrace`. In those cases should the attribute value be empty or should the attribute be omitted entirely?

Reply:
Java SDK doesn't care and sets an empty string. Should we do
```scala
if (exception.getStackTrace.nonEmpty) {
  ...
  
}
```
?

Reply:
Interesting. I guess `NoStackTrace` is a Scala-ism 😅 yeah, maybe better to omit in that case, I don't see the value.

Reply:
Sorry, on second look I'm not sure. I'm reading this page.

https://opentelemetry.io/docs/specs/otel/trace/exceptions/#attributes

> Additionally, the following attributes SHOULD be filled out:
> 
>  -  `exception.escaped`
>  -  `exception.message`
>  -  `exception.stacktrace`
>  -  `exception.type`

It's SHOULD not MUST. Also it says "filled out" and not "present", so I suspect we can't satisfy it either way if we don't have a stack trace. Are we missing the `escaped` attribute?

Reply:
More spec: https://opentelemetry.io/docs/specs/semconv/exceptions/exceptions-spans/

Reply:
Hm. 

>  `exception.escaped` - boolean - SHOULD be set to true if the exception event is recorded at a point where it is known that the exception is escaping the scope of the span. [1] - Recommended

I believe it's possible only with the manual instrumentation:

escaped:
```scala
Tracer[F].span("span").use { span =>
  io.onError { error =>
    span.recordException(new Exception, Attribute("exception.escaped", true))
  }
}
```

not escaped:
```scala
Tracer[F].span("span").use { span =>
  io.handleErrorWith { error =>
    span.recordException(new Exception, Attribute("exception.escaped", false))
  }
}
```

Reply:
> I believe it's possible only with the manual instrumentation:

I'm not caught up with the implementation details, but if `recordException` is called automatically can't we just check the `ExitCase` after the `use`?

Reply:
I guess we can find a way, for example, here: https://github.com/typelevel/otel4s/pull/325/files#diff-9bac43fa0f17631759132ba638deda5d102310637d050a1a6495d58fe44b96e2R119-R125

We can explore these options when we start working on the SdkSpanBuilder implementation.